genrl/core/actor_critic.py

```diff
@@ -144,13 +162,145 @@ def __init__(
         self.action_scale = kwargs["action_scale"] if "action_scale" in kwargs else 1
         self.action_bias = kwargs["action_bias"] if "action_bias" in kwargs else 0
 
+    def get_params(self):
+        actor_params = self.actor.parameters()
+        critic_params = list(self.critic1.parameters()) + list(
+            self.critic2.parameters()
+        )
+        return actor_params, critic_params
+
+    def forward(self, x):
+        q1_values = self.critic1(x).squeeze(-1)
+        q2_values = self.critic2(x).squeeze(-1)
+        return (q1_values, q2_values)
+
+    def get_action(self, state: torch.Tensor, deterministic: bool = False):
+        state = torch.as_tensor(state).float()
+
+        if self.actor.sac:
+            mean, log_std = self.actor(state)
+            std = log_std.exp()
+            distribution = Normal(mean, std)
+
+            action_probs = distribution.rsample()
+            log_probs = distribution.log_prob(action_probs)
+            action_probs = torch.tanh(action_probs)
+
+            action = action_probs * self.action_scale + self.action_bias
+
+            # enforcing action bound (appendix of SAC paper)
+            log_probs -= torch.log(
+                self.action_scale * (1 - action_probs.pow(2)) + np.finfo(np.float32).eps
+            )
+            log_probs = log_probs.sum(1, keepdim=True)
+            mean = torch.tanh(mean) * self.action_scale + self.action_bias
+
+            action = (action.float(), log_probs, mean)
+        else:
+            action = self.actor.get_action(state, deterministic=deterministic)
+
+        return action
+
+    def get_value(self, state: torch.Tensor, mode="first") -> torch.Tensor:
+        """Get Values from the Critic
+
+        Arg:
+            state (:obj:`torch.Tensor`): The state(s) being passed to the critics
+            mode (str): What values should be returned. Types:
+                "both" --> Both values will be returned
+                "min" --> The minimum of both values will be returned
+                "first" --> The value from the first critic only will be returned
+
+        Returns:
+            values (:obj:`list`): List of values as estimated by each individual critic
+        """
+        state = torch.as_tensor(state).float()
+
+        if mode == "both":
+            values = self.forward(state)
+        elif mode == "min":
+            values = self.forward(state)
+            values = torch.min(*values).squeeze(-1)
+        elif mode == "first":
+            values = self.critic1(state)
+        else:
+            raise KeyError("Mode doesn't exist")
+
+        return values
+
+
+class MlpSharedSingleActorMultiCritic(BaseActorCritic):
```

Comment:
Have this inherit from `MlpSingleActorMultiCritic` instead. A lot of code can be eliminated. You can use super().get_action(self.get_features(state)) or something instead. Code would be shortened a lot. Also please update the docstrings here.

Reply:
Currently the `actor_critic.py` file is way too big